arch/x86/compare256_sse2.c

```diff
@@ -0,0 +1,97 @@
+/* compare256_sse2.c -- SSE2 version of compare256
+ * Copyright Adam Stylinski <kungfujesus06@gmail.com>
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "../../zbuild.h"
+#include "../../zutil.h"
+
+#include "fallback_builtins.h"
+
+#if defined(X86_SSE2) && defined(HAVE_BUILTIN_CTZ)
+
+#include <emmintrin.h>
+
+static inline uint32_t compare256_unaligned_sse2_static(const uint8_t *src0, const uint8_t *src1) {
+    uint32_t len = 0;
+    int align_offset = ((uintptr_t)src0) & 15;
+    const uint8_t *end0 = src0 + 256;
+    const uint8_t *end1 = src1 + 256;
+    __m128i xmm_src0, xmm_src1, xmm_cmp;
+
+    /* Do the first load unaligned, than all subsequent ones we have at least
+     * one aligned load. Sadly aligning both loads is probably unrealistic */
+    xmm_src0 = _mm_loadu_si128((__m128i*)src0);
+    xmm_src1 = _mm_loadu_si128((__m128i*)src1);
+    xmm_cmp = _mm_cmpeq_epi8(xmm_src0, xmm_src1);
+
+    unsigned mask = (unsigned)_mm_movemask_epi8(xmm_cmp);
+
+    /* Compiler _may_ turn this branch into a ptest + movemask,
+     * since a lot of those uops are shared and fused */
+    if (mask != 0xFFFF) {
+        uint32_t match_byte = (uint32_t)__builtin_ctz(~mask);
+        return len + match_byte;
+    }
+
+    int align_adv = 16 - align_offset;
+    len += align_adv;
+    src0 += align_adv;
+    src1 += align_adv;
```

Comment:
Why is having both loads aligned unrealistic as mentioned above? Couldn't you calculate the alignment for `src1` different from `src0`? Then both `src0` and `src1` would be aligned?

Reply:
Obviously the strings would become different lengths in that case...

Reply:
@nmoinvaz Does this make sense?  Or did you have something else in mind?

Reply:
Yes that makes sense. Thank you!

Reply:
There is the possibility of shifting the strings over once they are in registers and combining them with other half, reconstituting what they'd be with a full unaligned load (similar to how you're _supposed_ to do this with altivec), but the shifts or "align" bit here is likely to be more expensive than what you'd gain.  Predominantly because you have a variable shift between 0 and 15 bytes determined at runtime, and the instruction for this requires an immediate, so it'd be a 15 entry jump table.